File: Web/pow.js
class Blockchain {
    constructor() {
        this.chain = [this.createGenesisBlock()];

        // 난이도 설정
        this.difficulty = 4;
    }

    // 제네시스 블록 생성
    createGenesisBlock() {
        return new Block(0, "01/01/2023", "Genesis Block", "0");
    }

    // 가장 최근 블록 반환
    getLatestBlock() {
        return this.chain[this.chain.length - 1];
    }

    // 새로운 블록 추가
    addBlock(newBlock) {
        // 이전 블록의 해시를 현재 블록의 previousHash로 설정
        newBlock.previousHash = this.getLatestBlock().hash;
        
        // 새로운 블록의 작업 증명(mine) 수행
        newBlock.mineBlock(this.difficulty);

        // 체인에 새로운 블록 추가
        this.chain.push(newBlock);
    }

    // 블록체인의 유효성 검증
    isChainValid() {
        for (let i = 1; i < this.chain.length; i++) {  
            const currentBlock = this.chain[i];
            const previousBlock = this.chain[i - 1];

            // 현재 블록의 해시 값이 유효한지 검증
            if (currentBlock.hash !== currentBlock.calculateHash()) {
                return false;
            }

            // 현재 블록의 이전 해시 값과 이전 블록의 해시 값이 일치하는지 검증
            if (currentBlock.previousHash !== previousBlock.hash) {
                return false;
            }
        }

        return true;
    }
}

File: Web/Pow.py
import hashlib

class Block:
	def __init__(self, data, previous_hash):
		"""
		Initializes a new Block object with
		the given data and previous hash.
		"""
		self.data = data
		self.previous_hash = previous_hash
		self.nonce = 0
		self.hash = self.calculate_hash()

	def calculate_hash(self):
		"""
		Calculates the SHA-256 hash of the
		block's data, previous hash, and nonce.
		"""
		sha = hashlib.sha256()
		sha.update(str(self.data).encode('utf-8') +
				str(self.previous_hash).encode('utf-8') +
				str(self.nonce).encode('utf-8'))
		return sha.hexdigest()

	def mine_block(self, difficulty):
		"""
		Mines the block using the Proof-of-Work algorithm
		with the given difficulty level.
		"""
		while self.hash[0:difficulty] != "0" * difficulty:
			self.nonce += 1
			self.hash = self.calculate_hash()

		print("Block mined:", self.hash)

class Blockchain:
	def __init__(self):
		"""
		Initializes a new Blockchain object with
		a genesis block.
		"""
		self.chain = [self.create_genesis_block()]

	def create_genesis_block(self):
		"""
		Creates the first block of the blockchain with
		arbitrary data and a previous hash of "0".
		"""
		return Block("Genesis Block", "0")

	def add_block(self, new_block):
		"""
		Adds a new block to the blockchain with
		the given data.
		"""
		new_block.previous_hash = self.chain[-1].hash
		
		# Difficulty level of 4
		new_block.mine_block(4)
		self.chain.append(new_block)

# create a new blockchain
blockchain = Blockchain()

# create and add three blocks to the blockchain
block1 = Block("Transaction data 1", "")
blockchain.add_block(block1)

block2 = Block("Transaction data 2", "")
blockchain.add_block(block2)

block3 = Block("Transaction data 3", "")
blockchain.add_block(block3)
